File: demo-bindings/python-bindings/conway/utility.py
#!/usr/bin/env python3
import json


class State(object):

    def __init__(self, line):
        jsondict = json.loads(line)
        self.field = list()
        for j in jsondict["field"]:
            self.field.append([k for k in j])

        self.rows = len(self.field)
        self.cols = len(self.field[0])

        self.maxColonisationDistance = jsondict["maxColonisationDistance"]
        self.cellsRemaining = jsondict["cellsRemaining"]
        self.cellGainPerTurn = jsondict["cellGainPerTurn"]
        self.maxGameIterations = jsondict["maxGameIterations"]
        self.maxCellCapacity = jsondict["maxCellCapacity"]
        self.currIteration = jsondict["currIteration"]

    def __getitem__(self, key):
        return self.field[key]

    def out_of_bounds(self, x, y):
        return x < 0 or y < 0 or x > self.cols or y > self.rows

    def is_my_cell(self, x, y):
        return not self.out_of_bounds(x, y) and self.field[x][y] == '#'

    def is_enemy_cell(self, x, y):
        return not self.out_of_bounds(x, y) and self.field[x][y] == 'O'

    def is_empty(self, x, y):
        return not self.out_of_bounds(x, y) and self.field[x][y] == '.'

    @staticmethod
    def submit_move(cell_list):
        print(json.dumps({"cells": cell_list}), end='\n', flush=True)

File: demo-bindings/cpp-bindings/utility.cpp
/*!
  utility.cpp provides a user with useful methods for JSON communication.
*/

#include <string>
#include <iostream>
#include <cmath>

#include "json/src/json.hpp"
#include "utility.hpp"
using nlohmann::json;

State::State(std::string json_line) {
    auto state = json::parse(json_line);

    maxColonisationDistance = state["maxColonisationDistance"];
    cellGainPerTurn = state["cellGainPerTurn"];
    cellsRemaining = state["cellsRemaining"];
    maxGameIterations = state["maxGameIterations"];
    maxCellCapacity = state["maxCellCapacity"];
    currIteration = state["currIteration"];
    field = state["field"].get<std::vector<std::string>>();
    rows = field.size();
    cols = field[0].size();
}

const bool State::inField(int row, int col) const {
  return row >= 0 && col >= 0 && row < rows && col < cols;
}

const bool State::isMineCell(int row, int col) const {
  return inField(row, col) && field[row][col] == '#';
}

const bool State::isEnemyCell(int row, int col) const {
  return  inField(row, col) && field[row][col] == 'O';
}

const bool State::isEmptyCell(int row, int col) const {
  return  inField(row, col) && field[row][col] == '.';
}

std::ostream& State::commitAction(std::ostream& stream, const std::vector<std::pair<int, int > >& cells) {
    json j = json::object();
    j["cells"] = json::array();

    for (auto c : cells) {
        j["cells"].push_back({c.first, c.second});
    }

    stream << j << std::endl;
    return stream;
}
